### ISSUE TITLE
Creating nine patch image backgrounds in StyleManager custom skin file 

### ISSUE THREAD

**Issue by iamchathu** [mgarin/weblaf#279]
Hi,

I really like custom skin feature in WebLook and Feel. I need change all my components like Futurico styling like demo.I have custom nine patch images for text fields and panel and I need to set them from custom skin.xml file.

Please guide me.

Thank you

**Comment by mgarin** [comment]
You might want to read `StyleManager` wiki first:
https://github.com/mgarin/weblaf/wiki/How-to-use-StyleManager
It generally describes how new styling system works.

Though, it isn't finished yet and I am working on it right now.
In current version there are a few blocking issues not allowing you to create custom skin with ease.

The best way to style any panel/field with nine-patch images is to provide a custom painter into that panel/field directly in the code. You can look at the futurico or android examples in demo app to find out how that could be done.

You can use skin right now, but there might be some issues.
Here is a small example:

**MyApp.java**
Simple application with styled panel.
```java
public class MyApp
{
    public static void main ( final String[] args )
    {
        StyleManager.setDefaultSkin ( MySkin.class );
        WebLookAndFeel.install ();
        TestFrame.show ( new WebPanel ( "my-panel" ).setPreferredSize ( 100, 100 ), 50 );
    }
}
```

**MySkin.java**
Your custom skin. You can also use `CustomSkin` class directly without creating a separate skin class, but in that case you won't be able to provide it as a default style which gives a performance boost at the application start (unnecessary default skin init will be skipped).
```java
public class MySkin extends CustomSkin
{
    public MySkin ()
    {
        super ( "resources/skin.xml" );
    }
}
```

**MyPanelPainter**
This one is required right now since all 9-patch resource paths will be relative to that class.
```java
public class MyPanelPainter<E extends JPanel> extends NPPanelPainter<E>
{
    // Made for the sake of simple relative paths
}
```

**skin.xml**
Base skin class. You can actually add all styles (label/panel/...) right here without creating separate XML files and including them, but I prefer separating styles for each specific component.
```xml
<skin>

    <!-- Skin information (not necessary) -->
    <id>my.ninepatch.skin</id>
    <name>My 9-patch skin</name>
    <description>Custom 9-patch-based UI skin</description>
    <author>Mikle Garin</author>
    <supportedSystems>all</supportedSystems>

    <!-- Base skin class, included resources with no specified "nearClass" attribute will use this class -->
    <class>my.project.MySkin</class>

    <!-- Separate style classes -->
    <include>resources/panel.xml</include>

</skin>
```

**panel.xml**
Panel style class. We have a default one here which is applied to all panels by default and custom "my-panel" styling that uses our custom panel painter.
```xml
<skin>

    <!-- JPanel styling -->
    <style type="panel" id="default">
        <component>
            <background>255,255,255</background>
        </component>
        <ui>
            <undecorated>true</undecorated>
        </ui>
    </style>

    <!-- My JPanel styling -->
    <style type="panel" id="my-panel" extends="default">
        <ui>
            <undecorated>false</undecorated>
        </ui>
        <painter id="painter" class="my.project.MyPanelPainter">
            <undecorated>false</undecorated>
            <backgroundIcon>resources/panel.9.png</backgroundIcon>
            <hiddenSideSpacing>3</hiddenSideSpacing>
        </painter>
    </style>

</skin>
```

So basically - you specify some custom panel styling under ID "my-panel" and simply provide that ID into panel you want to use that styling. Default panel styling is recommended to be undecorated since it is also applied to root pane, layered pane and some other elements you might not want to be styled at all.

**Comment by mgarin** [comment]
I will provide a complete wiki article on how custom skins can be created as soon as styling scheme is final. For now I just don't recommend using it as I am working on it right now and everything is a subject to change.

**Comment by iamchathu** [comment]
https://github.com/mgarin/weblaf/wiki/How-to-use-StyleManager

I read this article and i created my customskin.xml and loaded it the problem came to me use the background png to my TextFields

     <painter id="painter" class="my.project.MyPanelPainter">
            <undecorated>false</undecorated>
            <backgroundIcon>resources/panel.9.png</backgroundIcon>
            <hiddenSideSpacing>3</hiddenSideSpacing>
        </painter>

So I need to create custom PainterClass or can i use existing BackgroundPainter class like TexturePainter?

Can I know the location of resources/panel.9.png related to? (Is it related to MyCustomSkin.class)

**Comment by mgarin** [comment]
This is why I do not recommend using skin XML files yet :)
There will be too many questions on how this scheme works and some I simply cannot answer yet.

> I read this article and i created my customskin.xml and loaded it the problem came to me use the background png to my TextFields

Text fields aren't supported by the StyleManager yet, you can only specify their painters in code, not through skin XML file.

> So I need to create custom PainterClass or can i use existing BackgroundPainter class like TexturePainter?

You can use existing ones and can create your own as well.

> Can I know the location of resources/panel.9.png related to? (Is it related to MyCustomSkin.class)

Path is always related to the specified skin class. This is why it is not too convenient so far to work with some existing classes like nine-patch based ones - you will have to specify path relative to that class and that is a real issue. I will be improving this for sure.

**Comment by mgarin** [comment]
Also, to avoid a lot of confusion and questions - there will be a tool available to work with these style files effectively. Something like mini-IDE for this specific format. I will be releasing this tool together with  final StyleManager version.

**Comment by iamchathu** [comment]
I really Like this feature and waiting for it to complete. Can I know about the date please.

Currently I'm using 1.28

**Comment by iamchathu** [comment]
https://github.com/mgarin/weblaf/issues/279#issuecomment-101235278

I tried it and It's worked.But the problem is the image is stretched.Can I make it tiled? 

(Wow It's working :+1: )

**Comment by mgarin** [comment]
> I really Like this feature and waiting for it to complete. Can I know about the date please.

Unfortunately I can't provide it yet.
I was hoping to finish it long ago, but unfortunately got swallowed by other projects.

WebLaF is a semi-commercial project, yes, but unfortunately its commercial profit covers only small part of the time spent on development so I mostly work with it in my free time. Thus I cannot guarantee improvements and features release time since those usually require a lot of time.

> I tried it and It's worked.But the problem is the image is stretched.Can I make it tiled?

It should be stretched in case you are using 9-patch image and not a plain one.
And since you said before:

> I have custom nine patch images for text fields and panel and I need to set them from custom skin.xml file.

I assume you have 9-patch images, right?

**Comment by iamchathu** [comment]
>I have custom nine patch images for text fields and panel and I need to set them from custom skin.xml file.
>I assume you have 9-patch images, right?

Yes I have use nine patch editor and saved .9. files. So I how i can use them for Text Fields? 

Thank you for contributing to this project. It's the best UI solution I found for Java SE. 
You are doing great service to community.I really appreciate it.

I try to contribute this project.But I still trying to understand project yet.

**Comment by mgarin** [comment]
> Is there any work around to use non 9-patch image to make tiled background?

There is no workaround - application requires information about stretchable image areas, otherwise how could it know what part of the image should be stretched and which should not?

Though, you can create a `NinePatchIcon` from a plain image and specify stretchable areas and content margin manually if you know them, here is an example:
```java
final BufferedImage image = ImageUtils.getBufferedImage ( "C:\\image.png" );
final NinePatchIcon ninePatchIcon = NinePatchIcon.create ( image );
ninePatchIcon.addHorizontalStretch ( 0, 5, true );
ninePatchIcon.addHorizontalStretch ( 6, width - 6, false );
ninePatchIcon.addHorizontalStretch ( width - 5, width, true );
ninePatchIcon.addVerticalStretch ( 0, 5, true );
ninePatchIcon.addVerticalStretch ( 6, width - 6, false );
ninePatchIcon.addVerticalStretch ( width - 5, width, true );
ninePatchIcon.setMargin ( 10, 10, 10, 10 );
```
This will create a `NinePatchIcon` with [6;width-6] stretchable hor/ver areas and [11;width-11] content areas. You can use anywhere after this and it will be rendered properly.

This feature was added for the 9-patch files editor mostly, not for the manual usage - as you can see it isn't too convenient and simple.

> Yes I have use nine patch editor and saved .9. files. So I how i can use them for Text Fields?

Simply apply a custom painter to that textfield.
The best painter you can use for textfield right now is `NinePatchStatePainter`:
```java
final WebTextField textField = new WebTextField ( "Styled text field" );

final NinePatchStatePainter painter = new NinePatchStatePainter (  );
painter.addStateIcon ( ComponentState.normal, new NinePatchIcon ( "C:\\field.9.png" ) );
painter.addStateIcon ( ComponentState.focused, new NinePatchIcon ( "C:\\focused-field.9.png" ) );
textField.setPainter ( painter );
```
You can pass a manually created 9-patch icon there as well.

**Comment by mgarin** [comment]
You can also always apply single state painters like this:
```java
final WebTextField textField = new WebTextField ( "Styled text field" );
textField.setPainter ( new NinePatchIconPainter ( new NinePatchIcon ( "C:\\field.9.png" ) ) );
```
It is useful for panels or labels which usually have only one state.

**Comment by iamchathu** [comment]
Is there any guide about how to use resource file to define path of icons?

**Comment by mgarin** [comment]
There is none yet, it will be out with `StyleManager` release.

**Comment by mgarin** [comment]
There will be more options available to provide image resources with the next update.
Also a complete default skin for all components will be available there.
9-patch skin and related painters for various components will be added with the next updates.

**Comment by mgarin** [comment]
Just a short update on this request -

Some time ago `NinePatchDecoration` class have been added as a replacement for all custom 9-patch icon -based painters. It can now be used in all components to provide decoration for any specific component state based on the specified 9-patch icon.

Today I have pushed last changes allowing proper component styling using 9-patch resources. 
Here is, for example, how custom button styling works - first we need a style XML:
```xml
    <style type="button" id="my-button">
        <painter>
            <decorations overwrite="true">
                <ninepatch>
                    <icon>icons/normal.9.png</icon>
                </ninepatch>
                <ninepatch states="hover">
                    <icon>icons/hover.9.png</icon>
                </ninepatch>
                <ninepatch states="pressed">
                    <icon>icons/pressed.9.png</icon>
                </ninepatch>
            </decorations>
        </painter>
    </style>
```
You can put this one into your own custom skin or into the skin extension. Extensions were introduced only a few weeks ago and they are basically the same thing as skins, but you are limited to extending and creating new styles there, but you cannot override existing ones. That should be enough for this specific case anyway.

Also a small note - icon paths are specified related to according painter (`ButtonPainter` class in this case), but you can specify your own class like this:
```xml
    <ninepatch>
        <icon nearClass="some.package.ClassName">icons/normal.9.png</icon>
    </ninepatch>
```

Next - usage of button with your custom style within the code:
```java
new WebButton ( StyleId.of ( "my-button" ) );
```

That is pretty much all you should know to style any component. Of course different component have different states which could be decorated, but almost all components have a set of basic states like `enabled`,`disabled` and `focused`. Complex components like trees or tables simply contain multiple painters for each specific part of the component - for example nodes, rows, columns, selection etc.
